job_definitions/release_application_paas.yml

```diff
@@ -53,14 +53,33 @@
               }
 
               stage('Deploy') {
-                withEnv(["DM_CREDENTIALS_REPO=/var/lib/jenkins/digitalmarketplace-credentials", "CF_HOME=${pwd()}"]) {
-                  paas_credentials = sh(script: '$DM_CREDENTIALS_REPO/sops-wrapper -d $DM_CREDENTIALS_REPO/jenkins-vars/paas_credentials_env.enc', returnStdout: true).trim()
-                  withEnv(paas_credentials.tokenize("\n")) {
-                      sh('make paas-login')
+                  withEnv(["DM_CREDENTIALS_REPO=/var/lib/jenkins/digitalmarketplace-credentials", "CF_HOME=${pwd()}"]) {
+                    paas_credentials = sh(script: '$DM_CREDENTIALS_REPO/sops-wrapper -d $DM_CREDENTIALS_REPO/jenkins-vars/paas_credentials_env.enc', returnStdout: true).trim()
+                    withEnv(paas_credentials.tokenize("\n")) {
+                        sh('make paas-login')
+                    }
                   }
-                  lock("paas-${STAGE}-${APPLICATION_NAME}-deploy") {
-                      sh('make ${STAGE} deploy-app')
+
+                lock("paas-${STAGE}-${APPLICATION_NAME}-deploy") {
+                  if ("$RELEASE_NAME" == "") {
+                    echo "No release specified. Detecting currently deployed release version of ${APPLICATION_NAME} ..."
+
+                    /*
+                      This is a pretty painful command to parse. It essentially does the following:
+
+                      1) Hits `cf app ${APPLICATION_NAME} --guid` to get the CloudFoundry GUID of an application.
+                      2) Uses `cf curl /v2/apps/guid/routes` to work out the subdomain/path suffix an application is registered on.
+                      3) Uses `curl https://host.cloudapps.digital/path/_status` to retrieve the currently-released version of the app.
+
+                      We didn't find a way to do this across multiple lines, which would allow having intermediate variables to increase
+                      readability, due to how Jenkins escapes/injects variables between Groovy and the subshell.
+                    */
+                    RELEASE_NAME = sh(script: "curl -s https://\$(cf curl /v2/apps/\$(cf app ${APPLICATION_NAME} --guid)/routes | jq -crM '.resources[0].entity.host').cloudapps.digital\$(cf curl /v2/apps/\$(cf app ${APPLICATION_NAME} --guid)/routes | jq -crM '.resources[0].entity.path')/_status | jq -crM '.version'", returnStdout: true)
```

Comment:
George spotted this in the index alias script - multi line script within a single `sh()` that might work? https://github.com/alphagov/digitalmarketplace-jenkins/blob/master/job_definitions/update_index_alias.yml#L31

Reply:
👍